doc/sphinx/source/configuration.rst
.. _configuration:

Configuration
=============
You can copy the supplied sample-config to your home directory and move it to
``$XDG_CONFIG_HOME/bpython/config`` [#f1]_. bpython tries to find
``$XDG_CONFIG_HOME/bpython/config`` and use it as its configuration, if the
file does not exist bpython will use its documented defaults.

.. :: Footnotes

.. [#f1] ``$XDG_CONFIG_HOME`` defaults to ``~/.config`` if not set.

.. include:: configuration-options.rst


doc/sphinx/source/configuration-options.rst
General
-------
This refers to the ``[general]`` section in your `$XDG_CONFIG_HOME/bpython/config` file.

auto_display_list
^^^^^^^^^^^^^^^^^
Display the autocomplete list as you type (default: True).
When this is off, you can hit tab to see the suggestions.

autocomplete_mode
^^^^^^^^^^^^^^^^^
There are three modes for autocomplete. simple, substring, and fuzzy.
Simple matches methods with a common prefix, substring matches methods with a common
subsequence, and fuzzy matches methods with common characters (default:
simple).

.. versionadded:: 0.12

syntax
^^^^^^
Syntax highlighting as you type (default: True).

arg_spec
^^^^^^^^
Display the arg spec (list of arguments) for callables, when possible (default: True).

hist_file
^^^^^^^^^
History file (default: ``~/.pythonhist``).

paste_time
^^^^^^^^^^
The time between lines before pastemode is activated in seconds (default: 0.02).

hist_length
^^^^^^^^^^^
Number of lines to store in history (set to 0 to disable) (default: 100)

tab_length
^^^^^^^^^^
Soft tab size (default 4, see pep-8)

pastebin_url
^^^^^^^^^^^^
The pastebin url to post to (without a trailing slash). This pastebin has
to be a pastebin which uses LodgeIt. Examples are: http://paste.pocoo.org/xmlrpc/ and
http://bpaste.net/xmlrpc/ (default: http://bpaste.net/xmlrpc/)

pastebin_private
^^^^^^^^^^^^^^^^
If the pastebin supports a private option to make a random paste id, use it.
Default: True).

.. versionadded:: 0.12

pastebin_show_url
^^^^^^^^^^^^^^^^^
The url under which the new paste can be reached. ``$paste_id`` will be
replaced by the ID of the new paste. Examples are: http://bpaste.net/show/$paste_id/
and http://paste.pocoo.org/show/$paste_id/ (default: http://bpaste.net/show/$paste_id/)

pastebin_helper
^^^^^^^^^^^^^^^

The name of a helper executable that should perform pastebin upload on bpython's
behalf. If set, this overrides `pastebin_url`. It also overrides `pastebin_show_url`,
as the helper is expected to return the full URL to the pastebin as the first word of
its output. The data is supplied to the helper via STDIN.

An example helper program is ``pastebinit``, available for most systems. The
following helper program can be used to create `gists <http://gist.github.com>`_:

.. code-block:: python

  #!/usr/bin/env python

  import sys
  import urllib2
  import json

  def do_gist_json(s):
      """ Use json to post to github. """
      gist_public = False
      gist_url = 'https://api.github.com/gists'

      data = {'description': None,
              'public': None,
              'files' : {
                  'sample': { 'content': None }
              }}
      data['description'] = 'Gist from BPython'
      data['public'] = gist_public
      data['files']['sample']['content'] = s

      req = urllib2.Request(gist_url, json.dumps(data), {'Content-Type': 'application/json'})
      try:
          res = urllib2.urlopen(req)
      except HTTPError, e:
          return e

      try:
          json_res = json.loads(res.read())
          return json_res['html_url']
      except HTTPError, e:
          return e

  if __name__ == "__main__":
    s = sys.stdin.read()
    print do_gist_json(s)


.. versionadded:: 0.12

.. _configuration_color_scheme:

color_scheme
^^^^^^^^^^^^
See :ref:`themes` for more information.

Color schemes should be put in ``$XDG_CONFIG_HOME/bpython/``

e.g to use the theme ``$XDG_CONFIG_HOME/bpython/foo.theme`` set ``color_scheme = foo``

If you set the colorscheme to `foo` this will be translated to
``$XDG_CONFIG_HOME/bpython/foo.theme`` so be sure to put the file in that directory.

Leave blank or set to "default" to use the default (builtin) theme.

flush_output
^^^^^^^^^^^^
Whether to flush all output to stdout on exit (default: True).

save_append_py
^^^^^^^^^^^^^^
Whether to append ``.py`` to the filename while saving the input to a file.

.. versionadded:: 0.13

Keyboard
--------
This section refers to the ``[keyboard]`` section in your ``$XDG_CONFIG_HOME/bpython/config``.

You can set various keyboard shortcuts to be used by bpython. However, we have yet to map all keys
to their respective control codes. If you configure a key combination which is not yet supported
by bpython it will raise an exception telling you the key does not exist in bpython.keys.

Valid keys are:

* Control + any alphanumeric character (C-a through A-z, also a few others).
* Any function key ranging from F1 to F12.

pastebin
^^^^^^^^
Default: <F8>

last_output
^^^^^^^^^^^
Default: F9

Shows the last output in the systems $PAGER.

save
^^^^
Default: C-s

Saves the current session to a file (prompts for filename)

undo
^^^^
Default: C-r

Rewinds the last action.

up_one_line
^^^^^^^^^^^
Default: C-p

Move the cursor up, by one line.

down_one_line
^^^^^^^^^^^^^
Default: C-n

Move the cursor down, by one line.

cut_to_buffer
^^^^^^^^^^^^^
Default: C-k

Cuts the current line to the buffer.

search
^^^^^^
Default: C-o

Search up for any lines containing what is on the current line.

yank_from_buffer
^^^^^^^^^^^^^^^^
Default: C-y

Pastes the current line from the buffer (the one you previously cutted)

clear_word
^^^^^^^^^^
Default: C-w

Clear the word the cursor is currently on.

clear_line
^^^^^^^^^^
Default: C-u

Clears to the beginning of the line.

clear_screen
^^^^^^^^^^^^
Default: C-l

Clears the screen to the top.

show_source
^^^^^^^^^^^
Default: F2

Shows the source of the currently being completed (python) function.

exit
^^^^
Default: C-d

Exits bpython (use on empty line)

CLI
---
This refers to the ``[cli]`` section in your config file.

suggestion_width
^^^^^^^^^^^^^^^^
Default: 0.8

The width of the suggestion window in percent of the terminal width.

.. versionadded:: 0.9.8

trim_prompts
^^^^^^^^^^^^
Default: False

Trims lines starting with '>>> ' when set to True.

GTK
---
This refers to the ``[gtk]`` section in your `$XDG_CONFIG_HOME/bpython/config` file.

font
^^^^
Default: Monospace 10

The font to be used by the GTK version.


CHANGELOG
Changelog
=========

0.12
----

We want to give special thanks to the Hacker School project-
(https://www.hackerschool.com/) for choosing bpython as their pet hacking
project. In special we would like to thank the following people for contributing
their code to bpython:

- Martha Girdler
- Allison Kaptur
- Ingrid Cheung

We'd also like to thank Eike Hein for contributing his pastebin code which now
makes it possible to paste using a 3rd party program unlocking a whole slew of
pastebins for bpython users.

* Added a new pastebin_helper config option to name an executable that should
  perform pastebin upload on bpython's behalf. If set, this overrides
  pastebin_url. Data is supplied to the helper via STDIN, and it is expected
  to return a pastebin URL as the first word of its output.
* Fixed a bug causing pastebin upload to fail after a previous attempt was
  unsuccessful. A duplicate pastebin error would be displayed in this case,
  despite the original upload having failed.
* Added more key shortcuts to bpython.urwid
* Smarter dedenting after certain expressions
* #74 fixed broken completion when auto_display_list was disabled

We also have done numerous cleanup actions including building the man pages from
our documentation. Including the documentation in the source directory. Some
minor changes to the README to have EOL 79 and changes to urwid to work better
without twisted installed.

* Fix ungetch issues with Python 3.3. See issues #230, #231.

v0.11
-----

A bugfix/cleanup release .The fixed bugs are:

* #204: "import math" not autocompleting on python 3.2

Otherwise lots of small additions to the to be replacement for our ncurses
frontend, the urwid frontend.

I'd like to specifically thank Amjith Ramanujam for his work on history search
which was further implemented and is in working order right now.

v0.10.1
-------

A bugfix release. The fixed bugs are:

* #197: find_modules crashes on non-readable directories
* #198: Source tarball lacks .po files

v0.10
-----
As a highlight of the release, Michele Orrù added i18n support to bpython.

Some issues have been resolved as well:
* Config files are now located according to the XDG Base Directory
  Specification. The support for the old bpythonrc files has been
  dropped and ~/.bpython.ini as config file location is no longer supported.
  See issue #91.
* Fixed some issues with tuple unpacking in argspec. See issues #133 and #138.
* Fixed a crash with non-ascii filenames in import completion. See issue #139.
* Fixed a crash caused by inspect.findsource() raising an IndexError
  which happens in some situations. See issue #94.
* Non-ascii input should work now under Python 3.
* Issue #165: C-a and C-e do the right thing now in urwid.
* The short command-line option "-c config" was dropped as it conflicts with
  vanilla Python's "-c command" option. See issue #186.

v0.9.7.1
--------

A bugfix release. The fixed bugs are:

* #128: bpython-gtk is broken
* #134: crash when using pastebin and no active internet connection

v0.9.7
------

Well guys. It's been some time since the latest release, six months have passed
We have added a whole slew of new features, and closed a number of bugs as well.

We also have a new frontend for bpython. Marien Zwart contributed a urwid
frontend as an alternative for the curses frontend. Be aware that there still
is a lot to fix for this urwid frontend (a lot of the keyboard shortcuts do not
yet work for example) but please give it a good spin. Urwid also optionally
integrates with a Twisted reactor and through that with things like the GTK
event loop.

At the same time we have done a lot of work on the GTK frontend. The GTK
frontend is now 'usable'. Please give that a spin as well by running bpython-gtk
on you system.

We also welcome a new contributor in the name of Michele Orrù who we hope will
help us fix even more bugs and improve functionality.

As always, please submit any bugs you might find to our bugtracker.

* Pastebin confirmation added; we were getting a lot of people accidentally
  pastebinning sensitive information so I think this is a good idea.
* Don't read PYTHONSTARTUP when executed with -i.
* BPDB was merged in. BPDB is an extension to PDB which allows you to press B
  in a PDB session which will let you be dropped into a bpython sessions with
  the current PDB locals(). For usage, see the documentation.
* The clear word shortcut (default: C-w) now deletes to the buffer.
* More tests have been added to bpython.
* The pastebin now checks for a previous paste (during the session) with the
  exact same content to guard against twitchy fingers pastebinning multiple
  times.
* Let import completion return "import " instead of "import".

* GTK now has pastebin, both for full log as well as the current selection.
* GTK now has write2file.
* GTK now has a menu.
* GTK now has a statusbar.
* GTK now has show source functionality.
* GTK saves the pastebin url to the clipboard.
* GTK now has it's own configuration section.
* Set focus to the GTK text widget to allow for easier embedding in PIDA and
  others which fixes issues #121.

* #87: Add a closed attribute to Repl to fix mercurial.ui.ui expecting stderr
  to have this attribute.
* #108: Unicode characters in docsrting crash bpython
* #118: Load_theme is not defined.
* #99: Configurable font now documented.
* #123: <F8> Pastebin can't handle 'ESC' key
* #124: Unwanted input when using <arrow>/<FXX> keys in the statusbar prompt.


v0.9.6.2
--------
Unfortunately another bugfix release as I (Bob) broke py3 support.

* #84: bpython doesn't work with Python 3
       Thanks very much to Henry Prêcheur for both the bug report and the
       patch.

v0.9.6.1
--------
A quick bugfix release (this should not become a habit).

* #82: Crash on saving file.

v0.9.6
------
A bugfix/feature release (and a start at gtk). Happy Christmas everyone!

* #67: Make pastebin URL really configurable.

bpython/config.py
from __future__ import absolute_import
from __future__ import with_statement

import os
import sys

try:
    from configparser import ConfigParser
except ImportError:
    from ConfigParser import ConfigParser

from itertools import chain

from bpython.keys import cli_key_dispatch as key_dispatch
from bpython.autocomplete import SIMPLE as default_completion

MAGIC_METHODS = ', '.join('__%s__' % s for s in [
    'init', 'repr', 'str', 'lt', 'le', 'eq', 'ne', 'gt', 'ge', 'cmp', 'hash',
    'nonzero', 'unicode', 'getattr', 'setattr', 'get', 'set', 'call', 'len',
    'getitem', 'setitem', 'iter', 'reversed', 'contains', 'add', 'sub', 'mul',
    'floordiv', 'mod', 'divmod', 'pow', 'lshift', 'rshift', 'and', 'xor', 'or',
    'div', 'truediv', 'neg', 'pos', 'abs', 'invert', 'complex', 'int', 'float',
    'oct', 'hex', 'index', 'coerce', 'enter', 'exit']
)


class Struct(object):

    """Simple class for instantiating objects we can add arbitrary attributes
    to and use for various arbitrary things."""


def get_config_home():
    """Returns the base directory for bpython's configuration files."""
    xdg_config_home = os.environ.get('XDG_CONFIG_HOME', '~/.config')
    return os.path.join(xdg_config_home, 'bpython')


def default_config_path():
    """Returns bpython's default configuration file path."""
    return os.path.join(get_config_home(), 'config')


def fill_config_with_default_values(config, default_values):
    for section in default_values.keys():
        if not config.has_section(section):
            config.add_section(section)

        for (opt, val) in default_values[section].items():
            if not config.has_option(section, opt):
                config.set(section, opt, str(val))


def loadini(struct, configfile):
    """Loads .ini configuration file and stores its values in struct."""

    config_path = os.path.expanduser(configfile)
    if not os.path.isfile(config_path) and configfile == default_config_path():
        # We decided that '~/.bpython/config' still was a crappy
        # place, use XDG Base Directory Specification instead.  Fall
        # back to old config, though.
        config_path = os.path.expanduser('~/.bpython/config')

    config = ConfigParser()
    fill_config_with_default_values(config, {
        'general': {
            'arg_spec': True,
            'auto_display_list': True,
            'color_scheme': 'default',
            'complete_magic_methods': True,
            'magic_methods': MAGIC_METHODS,
            'autocomplete_mode': default_completion,
            'dedent_after': 1,
            'flush_output': True,
            'highlight_show_source': True,
            'hist_file': '~/.pythonhist',
            'hist_length': 100,
            'hist_duplicates': True,
            'paste_time': 0.02,
            'syntax': True,
            'tab_length': 4,
            'pastebin_confirm': True,
            'pastebin_private': False,
            'pastebin_url': 'http://bpaste.net/xmlrpc/',
            'pastebin_private': True,
            'pastebin_show_url': 'http://bpaste.net/show/$paste_id/',
            'pastebin_helper': '',
            'save_append_py': False
        },
        'keyboard': {
            'clear_line': 'C-u',
            'clear_screen': 'C-l',
            'clear_word': 'C-w',
            'cut_to_buffer': 'C-k',
            'delete': 'C-d',
            'down_one_line': 'C-n',
            'exit': '',
            'last_output': 'F9',
            'pastebin': 'F8',
            'save': 'C-s',
            'show_source': 'F2',
            'suspend': 'C-z',
            'undo': 'C-r',
            'search': 'C-o',
            'up_one_line': 'C-p',
            'yank_from_buffer': 'C-y'},
        'cli': {
            'suggestion_width': 0.8,
            'trim_prompts': False,
        },
        'gtk': {
            'font': 'monospace 10',
            'color_scheme': 'default'}})
    if not config.read(config_path):
        # No config file. If the user has it in the old place then complain
        if os.path.isfile(os.path.expanduser('~/.bpython.ini')):
            sys.stderr.write('Error: It seems that you have a config file at '
                             '~/.bpython.ini. Please move your config file to '
                             '%s\n' % default_config_path())
            sys.exit(1)

    struct.dedent_after = config.getint('general', 'dedent_after')
    struct.tab_length = config.getint('general', 'tab_length')
    struct.auto_display_list = config.getboolean('general',
                                                 'auto_display_list')
    struct.syntax = config.getboolean('general', 'syntax')
    struct.arg_spec = config.getboolean('general', 'arg_spec')
    struct.paste_time = config.getfloat('general', 'paste_time')
    struct.highlight_show_source = config.getboolean('general',
                                                     'highlight_show_source')
    struct.hist_file = config.get('general', 'hist_file')
    struct.hist_length = config.getint('general', 'hist_length')
    struct.hist_duplicates = config.getboolean('general', 'hist_duplicates')
    struct.flush_output = config.getboolean('general', 'flush_output')
    struct.pastebin_key = config.get('keyboard', 'pastebin')
    struct.save_key = config.get('keyboard', 'save')
    struct.search_key = config.get('keyboard', 'search')
    struct.show_source_key = config.get('keyboard', 'show_source')
    struct.suspend_key = config.get('keyboard', 'suspend')
    struct.undo_key = config.get('keyboard', 'undo')
    struct.up_one_line_key = config.get('keyboard', 'up_one_line')
    struct.down_one_line_key = config.get('keyboard', 'down_one_line')
    struct.cut_to_buffer_key = config.get('keyboard', 'cut_to_buffer')
    struct.yank_from_buffer_key = config.get('keyboard', 'yank_from_buffer')
    struct.clear_word_key = config.get('keyboard', 'clear_word')
    struct.clear_line_key = config.get('keyboard', 'clear_line')
    struct.clear_screen_key = config.get('keyboard', 'clear_screen')
    struct.delete_key = config.get('keyboard', 'delete')
    struct.exit_key = config.get('keyboard', 'exit')
    struct.last_output_key = config.get('keyboard', 'last_output')

    struct.pastebin_confirm = config.getboolean('general', 'pastebin_confirm')
    struct.pastebin_private = config.getboolean('general', 'pastebin_private')
    struct.pastebin_url = config.get('general', 'pastebin_url')
    struct.pastebin_private = config.get('general', 'pastebin_private')
    struct.pastebin_show_url = config.get('general', 'pastebin_show_url')
    struct.pastebin_helper = config.get('general', 'pastebin_helper')

    struct.cli_suggestion_width = config.getfloat('cli',
                                                  'suggestion_width')
    struct.cli_trim_prompts = config.getboolean('cli',
                                                'trim_prompts')

    struct.complete_magic_methods = config.getboolean('general',
                                                      'complete_magic_methods')
    methods = config.get('general', 'magic_methods')
    struct.magic_methods = [meth.strip() for meth in methods.split(',')]
    struct.autocomplete_mode = config.get('general', 'autocomplete_mode')
    struct.save_append_py = config.getboolean('general', 'save_append_py')

    struct.gtk_font = config.get('gtk', 'font')

    color_scheme_name = config.get('general', 'color_scheme')
    color_gtk_scheme_name = config.get('gtk', 'color_scheme')

    default_colors = {
        'keyword': 'y',
        'name': 'c',
        'comment': 'b',
        'string': 'm',
        'error': 'r',
        'number': 'G',
        'operator': 'Y',
        'punctuation': 'y',
        'token': 'C',
        'background': 'd',
        'output': 'w',
        'main': 'c',
        'paren': 'R',
        'prompt': 'c',
        'prompt_more': 'g',
    }

    default_gtk_colors = {
        'keyword': 'b',
        'name': 'k',
        'comment': 'b',
        'string': 'm',
        'error': 'r',
        'number': 'G',
        'operator': 'B',
        'punctuation': 'g',
        'token': 'C',
        'background': 'w',
        'output': 'k',
        'main': 'c',
        'paren': 'R',
        'prompt': 'b',
        'prompt_more': 'g',
    }

    if color_scheme_name == 'default':
        struct.color_scheme = default_colors
    else:
        struct.color_scheme = dict()

        theme_filename = color_scheme_name + '.theme'
        path = os.path.expanduser(os.path.join(get_config_home(),
                                               theme_filename))
        old_path = os.path.expanduser(os.path.join('~/.bpython',
                                                   theme_filename))

        for path in [path, old_path]:
            try:
                load_theme(struct, path, struct.color_scheme, default_colors)
            except EnvironmentError:
                continue
            else:
                break
        else:
            sys.stderr.write("Could not load theme '%s'.\n" %
                            (color_scheme_name, ))
            sys.exit(1)

    if color_gtk_scheme_name == 'default':
        struct.color_gtk_scheme = default_gtk_colors
    else:
        struct.color_gtk_scheme = dict()
        # Note: This is a new config option, hence we don't have a
        # fallback directory.
        path = os.path.expanduser(os.path.join(get_config_home(),
                                               color_gtk_scheme_name + '.theme'))

        try:
            load_theme(struct, path, struct.color_gtk_scheme, default_colors)
        except EnvironmentError:
            sys.stderr.write("Could not load gtk theme '%s'.\n" %
                            (color_gtk_scheme_name, ))
            sys.exit(1)

    # checks for valid key configuration this part still sucks
    for key in (struct.pastebin_key, struct.save_key):
        key_dispatch[key]


def load_theme(struct, path, colors, default_colors):
    theme = ConfigParser()
    with open(path, 'r') as f:
        theme.readfp(f)
    for k, v in chain(theme.items('syntax'), theme.items('interface')):
        if theme.has_option('syntax', k):
            colors[k] = theme.get('syntax', k)
        else:
            colors[k] = theme.get('interface', k)

    # Check against default theme to see if all values are defined
    for k, v in default_colors.items():
        if k not in colors:
            colors[k] = v
    f.close()
